Override user-defined pthread types

fbrausse opened the pull request:
This PR (*edit:* ~is based on #819, is work-in-progress atm. and~ not anymore) does:
- add support for a new type annotation `__attribute__((annotate("__ESBMC_ODR-override")))`
- add these annotations to all pthread-specific types in the c2goto models (this required changing some of them to become structures (structs or unions) instead of plain integers - the frontend ignores typedefs!)
- keep structure types even for variables symbolic until after the linker
- modifies the linker to prefer type symbols that are annotated with ODR-override even if an equally named symbol is already defined and present in the context; the preferred symbol retains its ODR-override annotation -> preference independent of link order

I intend to put this ODR-override behaviour of the linker behind a flag to explicitely opt into the above behaviour resolving [ODR](https://en.cppreference.com/w/cpp/language/definition) violations. Actually, ODR is a C++ specific concept, the analogue constraints from C would be "compatible type", which is weaker. Still, I chose to refer to ODR since "compatibility" between types to be overridden/replaced is nowhere defined or used in this implementation.

The purpose of this PR is to support the "weird SV-COMP pthread types". From https://github.com/esbmc/esbmc/pull/819#issuecomment-1190521394:

> Since https://github.com/esbmc/esbmc/pull/602 and https://github.com/esbmc/esbmc/pull/615 ESBMC has had problems with SV-COMP concurrency benchmarks that are caused by their "weird pthread types" in pre-processed sources, cf. https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1359. What I'm trying to get to here is adding an annotation to c2goto-lib's types that "override" SV-COMP / user-defined ones in order to ignore their nonsense. But for that we need correct information on all the types from all the TUs we're going to link. In a next PR (working on it), I wanted to enable the linker to recognize types (like pthread_mutex_t) with such an annotation and prefer that one over whatever else exists in the target context.

fbrausse committed:
[pthread] annotate <pthread.h> types with "__ESBMC_ODR-override"

This also changes things like pthread_t and pthread_key_t which have been
typedefs of plain integers to unions in order to attach the attribute to
those unions. Typedefs are being ignored by the frontend and for integers
no new symbols the linker could care about are added.

fbrausse committed:
[clang-c-frontend] make get_decl() implementation a bit more uniform among the cases

fbrausse committed:
[clang-c-frontend] move annotation processing to a separate function

fbrausse committed:
[c_link] cleanup cases in duplicate_*() and take "incomplete_union" into account

fbrausse committed:
[irep] add Boolean "odr_override" attribute (not used, yet)

fbrausse committed:
[clang-c-frontend] interpret "__ESBMC_ODR-override" annotation on structure type declarations

fbrausse committed:
[c_link] instead of renaming a new symbol of differing type, overwrite it when instructed by odr_override

fbrausse committed:
[c_link] collect info about symbols that need fixing

fbrausse committed:
[symex] argument passing: simulate conversion of non-matching parameters via constant union

fbrausse committed:
[c_link] fix: follow symbolic types in namespace

fbrausse committed:
[c2goto-libs] guard definition of pthread_attr_t to accomodate glibc

Glibc's <bits/types/sigevent_t.h> checks for this guard and if not present
(re-)defines the typedef incompatibly (as a union).

fbrausse committed:
[clang-c-frontend] simplify

fbrausse commented:
[Master](https://github.com/esbmc/esbmc/actions/runs/2679770255) (same as in https://github.com/esbmc/esbmc/pull/796#issuecomment-1187167842):
```
Statistics:          15648 Files
  correct:            8442
    correct true:     5118
    correct false:    3324
  incorrect:            56
    incorrect true:     25
    incorrect false:    31
  unknown:            7150
  Score:             12264 (max: 25567)
```
Earlier version 3de6b1eefe7a2245947b6fbf72334707df445187 of [this PR](https://github.com/esbmc/esbmc/actions/runs/2717420925):
```
Statistics:          15648 Files
  correct:            8428
    correct true:     5117
    correct false:    3311
  incorrect:            55
    incorrect true:     26
    incorrect false:    29
  unknown:            7165
  Score:             12249 (max: 25567)
```
Diff:
- [termination](https://fbrausse.github.io/esbmc/svcomp/diff-gh-796-master-gh-823/termination/results.2022-07-23_14-21-39.diff.html)
- [unreach-call](https://fbrausse.github.io/esbmc/svcomp/diff-gh-796-master-gh-823/unreach-call/results.2022-07-23_14-21-41.diff.html)
- [valid-memsafety](https://fbrausse.github.io/esbmc/svcomp/diff-gh-796-master-gh-823/valid-memsafety/results.2022-07-23_14-21-51.diff.html)

fbrausse committed:
[c_link] merged namespace prefers ODR-override

mikhailramalho reviewed (commented):
I saw that you went ahead with the incomplete struct/union, is that required for this to work?

Did you try setting the type as symbols and let the code before goto level expand them?

fbrausse commented:
> I saw that you went ahead with the incomplete struct/union, is that required for this to work?

That's a good question, I would need to double-check. I based this PR on #819 as this made it easier for me reason about the types `struct` and `union` get converted to. These "butchered" (not explicitly incomplete) types before it were confusing the linker.

> Did you try setting the type as symbols and let the code before goto level expand them?

Yes, for symbols introduced by `get_var` the type remains symbolic, same when `complete=false` during clang_c_convert. Some of the stuff failing here is due to the boolector backend not correctly handling the array API. It just up-casts `smt_astt` to its own AST node type, but actually it is an `array_ast`. This happens for updates of arrays of unions.

fbrausse committed:
[smt/array] fix wrong application of array_downcast

array_ite() wrongly assumed that ITE of array-members would result in array
AST nodes.

Fixes:
- crash on sv-benchmarks/c/aws-c-common/aws_hash_table_create_harness.i
  (result is still wrong)
- regression/esbmc-unix2/23_picosat-846_05
- regression/nonz3/29_exStbHwAcc

fbrausse committed:
[expr2c] do not assume init_field of constant_union is set

This is still TODO in smt_convt::get_by_ast()

fbrausse committed:
[regression] fix some wrong pthread API usage

fbrausse committed:
[clang-c-frontend] fix: gen_zero() does not cope with symbolic types

Alternatively, c_link as the only consumer of .zero_initializer could
perform the type resolution, though that might introduce ambiguities due
to the order in which symbol linking is performed

fbrausse committed:
[regression] fix more wrong pthread API usage

fbrausse committed:
clang-format

fbrausse commented:
> The following tests FAILED:
	 ~61 - regression/esbmc-unix/00_race18 (Failed)~
	 ~62 - regression/esbmc-unix/00_race19 (Failed)~
	 ~63 - regression/esbmc-unix/00_race20 (Failed)~

These are not directly related to this PR.
```c
pthread_t id[2]; /* pthread_t is a typedef of an anonymous union with a single member */
// [...]
pthread_create(&id[0], /* ... */);
```
fails in smt_conv in `convert_assign` due to mismatching smt AST node types: side1 is a `btor_smt_ast` and side2 is a `array_ast`.

>	697 - regression/esbmc/github_839 (Failed)

Uses the old frontend and already fails on master with `PARSING ERROR`.

>	1698 - regression/esbmc-cpp/cpp/cpp_with_pthread_01 (Failed)
	1700 - regression/esbmc-cpp/cpp/cpp_with_pthread_02 (Failed)
	~1701 - regression/esbmc-cpp/cpp/cpp_with_pthread_02-no-lib (Failed)~

These all fail in `clang_cpp_convertert::get_function()` with
```
ERROR: Declaration has an empty name:
CXXConstructorDecl 0x55555699bd38 </tmp/esbmc.fcd7-6224-ec77/headers/pthread.h:36:9> col:9 implicit used  'void () throw()' inline default trivial
`-CompoundStmt 0x55555699bff8 <col:9>
```
due to the union pthread_t gets typedef'ed to being anonymous.

However, SV-COMP results are not great, yet:

[Master](https://github.com/esbmc/esbmc/actions/runs/2992941184) ff1b9dfb6776ad7cb3e6e593b512885ff90007a3:
```
Statistics:          15648 Files
  correct:            8677
    correct true:     5180
    correct false:    3497
  incorrect:            64
    incorrect true:     26
    incorrect false:    38
  unknown:            6907
  Score:             12417 (max: 25567)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/2992944177) 689c829db040bbd2bc2bc099d8dfd6988de32198:
```
Statistics:          15648 Files
  correct:            8279
    correct true:     5069
    correct false:    3210
  incorrect:            65
    incorrect true:     27
    incorrect false:    38
  unknown:            7304
  Score:             11876 (max: 25567)
```
The additional incorrect one crashed before. I started looking into the new (correct -> unknown) ones and it seems that the context is still not fixed for the override types, in particular for statically initialized symbols e.g. in <https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/pthread/lazy01.i>.

- [no-overflow](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-gh-823-2/no-overflow/results.2022-09-08_10-41-49.diff.html)
- [termination](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-gh-823-2/termination/results.2022-09-08_10-41-50.diff.html)
- [unreach-call](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-gh-823-2/unreach-call/results.2022-09-08_10-41-51.diff.html)
- [valid-memsafety](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-gh-823-2/valid-memsafety/results.2022-09-08_10-42-00.diff.html)

fbrausse committed:
[symex] more log_{debug => status}() on dump()

fbrausse committed:
[clang-c] handle CXXRecord decls as well

Fixes regression/esbmc-cpp/cpp/cpp_with_pthread_02-no-lib

fbrausse committed:
[c_link] overwrite values of zero-initialized symbols

fbrausse committed:
[context] remove cruft

fbrausse committed:
[regression] fixed esbmc-unix/05_pfscan-1.0_01

fbrausse committed:
[pthread] use struct instead of union for single-component types

fbrausse committed:
[pthread] remove unused members of pthread_{mutex,cond}_t

fbrausse committed:
[regression] floats-regression/nextafter: assert equality w/o implicit type conversion

fbrausse commented:
I understand the failing Linux regression tests (see comment above), but I cannot reproduce the macOS ones:
> The following tests FAILED:
	705 - regression/esbmc/math_cos01 (Failed)
	1277 - regression/floats-regression/fma (Failed)
	1294 - regression/floats-regression/nextafter (Failed)

lucasccordeiro commented:
Rafael has a MacOS in the lab.

Can you speak with Rafael to reproduce this error?



Em sex., 9 de set. de 2022 às 13:44, Franz Brauße ***@***.***>
escreveu:

> I understand the failing Linux regression tests (see comment above), but I
> cannot reproduce the macOS ones:
>
> The following tests FAILED:
> 705 - regression/esbmc/math_cos01 (Failed)
> 1277 - regression/floats-regression/fma (Failed)
> 1294 - regression/floats-regression/nextafter (Failed)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/823#issuecomment-1241932142>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4F4HKYTK36ZWOA7JCYLNLV5MWLHANCNFSM54NV57PA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Lucas C. Cordeiro​

https://ssvlab.github.io/lucasccordeiro/

fbrausse committed:
[pthread] declare symbols only used internally symbols static

fbrausse committed:
[c_link] WIP: hack to zero-init pthread_mutex_t symbols in static context

fbrausse committed:
[expr] generalize is_zero() for exprs of compound type

fbrausse committed:
[c_link] guard the hack overwriting statically initialized symbols

fbrausse committed:
[dereference] fix another instance of building ref from wrong type

Fixes the "Unexpectedly array type'd argument to construct_struct_ref" in
SV-COMP's c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--mtd--mtd.ko-entry_point.cil.out.i
though that one still fails due to Boolector complaining about having to deal
with more than 134217724 variables.

fbrausse committed:
[c_link] hide debug output on standard verbosity

fbrausse committed:
[c_link] extend hack to init static pthread symbols to all that have static initialization defined

fbrausse commented:
[Master](https://github.com/esbmc/esbmc/actions/runs/3060937710):
```
Statistics:          15648 Files
  correct:            8551
    correct true:     5163
    correct false:    3388
  incorrect:            56
    incorrect true:     25
    incorrect false:    31
  unknown:            7041
  Score:             12418 (max: 25567)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/3045812379):
```
Statistics:          15648 Files
  correct:            8558
    correct true:     5174
    correct false:    3384
  incorrect:            57
    incorrect true:     26
    incorrect false:    31
  unknown:            7033
  Score:             12404 (max: 25567)
```
New wrong was unknown before. Diff:
- [termination](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-2-gh-823-4/termination/results.2022-09-17_13-47-52.diff.html)
- [unreach-call](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-2-gh-823-4/unreach-call/results.2022-09-17_13-47-53.diff.html)
- [valid-memsafety](https://fbrausse.github.io/esbmc/svcomp/diff-gh-823-master-2-gh-823-4/valid-memsafety/results.2022-09-17_13-48-01.diff.html)

fbrausse committed:
[clang-c-frontend] keep implicitely decayed parameter types

Fixes sv-benchmarks/c/ldv-regression/callfpointer.i, which failed to
recognize the call to g() as (*g)() in

  void f(void g(int)) {
   g(1);
  }

The original type according to Clang is the non-pointer function type,
which of course is not meant to be that of 'g'.

fbrausse committed:
[svcomp wrapper] add "--dry-run" flag

fbrausse committed:
[smt] more log_{debug => status} on dump()

fbrausse committed:
[c_link] rm debug output

fbrausse committed:
keep more types symbolic

Together with __ESBMC_ODR-override, this allows the user code to specify
pthread-types that will get replaced by the overridden ones from our c2goto
models.

In particular, the new regression test defines pthread_t and pthread_mutex_t
incompatibly but with size equal or larger to the model's ones. This is still
required! However, all the used pthread_*() functions as well as the mutex and
the thread array now correspond to the model's.

Notably, the constant PTHREAD_MUTEX_INITIALIZER does not work, yet, as the
clang-c-frontend assigns all (named) members (with zero). So the initializer
would be wrong and at this stage there is nothing much we can do as the
ODR-override from the c2goto libs is not even loaded, yet.

fbrausse committed:
[exec-state] for some reason c:pthread_lib.c@__ESBMC_num_threads_running must be considered global while others in pthread-lib.c do not have to

fbrausse committed:
cleanup

fbrausse committed:
cleanup execution_state

fbrausse committed:
[clang-c-frontend] remove debugging output

fbrausse committed:
cleanup

fbrausse committed:
[c_link] cleanup

fbrausse committed:
clang-format

fbrausse commented:
I've updated this development to current master. Fixes #1011.

fbrausse committed:
[cpp] fix dangling reference to type returned by ns.follow()

fbrausse committed:
[om] extend 'extern "C"' around all type definitions

fbrausse committed:
[regression] clang-cpp-frontend does not handle constructors of unnamed types

fbrausse commented:
[Master](https://github.com/esbmc/esbmc/actions/runs/5012739408):
```
Statistics:          26459 Files
  correct:           14434
    correct true:     8232
    correct false:    6202
  incorrect:            85
    incorrect true:     39
    incorrect false:    46
  unknown:           11940
  Score:             20682 (max: 43043)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/5034830729):
```
Statistics:          26459 Files
  correct:           14089
    correct true:     7854
    correct false:    6235
  incorrect:            90
    incorrect true:     42
    incorrect false:    48
  unknown:           12280
  Score:             19831 (max: 43043)
```
Summary of the * -> wrong:
- no-overflow: 2x timeout -> wrong: no change, just faster
- unreach-call: 3x unknown -> wrong: all were accessing the non-existant `__lock` member in `pthread_mutex_t`

There are no wrong -> *. All in all there's quite some fluctuation: previously fast ones now time out and previously unknown ones are now solved (correct and wrong as well, see above).

lucasccordeiro commented:
@fbrausse: can I ask you to run this PR over the concurrency category of SV-COMP?

fbrausse commented:
> @fbrausse: can I ask you to run this PR over the concurrency category of SV-COMP?

@lucasccordeiro Is it the "Run Benchexec (Task)" Github action with ConcurrencySafety-Main?

lucasccordeiro commented:
Yes :-)

fbrausse commented:
> @fbrausse: can I ask you to run this PR over the concurrency category of SV-COMP?

[Master](https://github.com/esbmc/esbmc/actions/runs/5046400200/jobs/9052253062):
```
Statistics:            689 Files
  correct:             373
    correct true:      107
    correct false:     266
  incorrect:             3
    incorrect true:      3
    incorrect false:     0
  unknown:             313
  Score:               384 (max: 1036)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/5043773666/jobs/9046625561):
```

Statistics:            689 Files
  correct:             375
    correct true:      107
    correct false:     268
  incorrect:             3
    incorrect true:      3
    incorrect false:     0
  unknown:             311
  Score:               386 (max: 1036)
```
